File: python/say/say.py
rows = {
    "ones": ["","one","two","three","four","five","six","seven","eight","nine"],
    "tens": ["ten","eleven","twelve","thirteen","fourteen","fifteen","sixteen","seventeen","eighteen","nineteen"],
    "twenties": ["","","twenty","thirty","forty","fifty","sixty","seventy","eighty","ninety"]
}

def say(number):
    if number < 0 or number >= 1_000_000_000_000:
        raise ValueError("input out of range")
    if number == 0:
        return 'zero'

    list_number = [*str(number)]
    num_digits = len(list_number)
    groups = [('billion', 9), ('million', 6), ('thousand', 3), ('', 0)]
    out = ''
    for g in groups:
        text, count = g[0], g[1]
        if num_digits > count:
            num = ''
            num += list_number[-(count + 3)] if num_digits > count + 2 else ''
            num += list_number[-(count + 2)] if num_digits > count + 1 else ''
            num += list_number[-(count + 1)]
            ninety_nine = to_ninety_nine(num)
            if ninety_nine:
                out += ' ' if out else ''
                out += ninety_nine
                if text != '': out += ' ' + text
    return out


def to_ninety_nine(number):
    list_number = [*str(number)]
    num_digits = len(list_number)
    ones = int(list_number[-1]) if num_digits > 0 else 0
    tens = int(list_number[-2]) if num_digits > 1 else 0
    hundreds = int(list_number[-3]) if num_digits > 2 else 0

    out = ''
    if hundreds > 0:
        out = rows["ones"][hundreds] + ' hundred'
    if tens > 1:
        out += ' ' if out else ''
        out += rows["twenties"][tens]
    elif tens == 1:
        out += ' ' if out else ''
        out += rows["tens"][ones]

    if ones > 0 and tens != 1:
        if tens > 1: out += "-"
        out += rows["ones"][ones]

    return out

File: ruby/say/say.rb
class Say
  DIGITS = %w[zero one two three four five six seven eight nine]
  TEENS = %w[ten eleven twelve thirteen fourteen fifteen sixteen seventeen eighteen nineteen]
  TENS = %w[zero ten twenty thirty forty fifty sixty seventy eighty ninety]
  ORDERS = {
    1_000_000_000 => "billion",
    1_000_000     => "million",
    1000          => "thousand",
    100           => "hundred",
  }

  def initialize(number)
    @number = number
  end

  def in_english
    throw ArgumentError if @number > 999_999_999_999 or @number < 0
    return "zero" if @number == 0
    out = []
    ORDERS.each_with_index { |(order, label), i|
      cnt = @number / order
      has_100 = false
      if cnt >= 100
        hundreds = cnt / 100
        out << up_to_100(hundreds) + " hundred"
        cnt -= hundreds * 100
        @number -= hundreds * 100 * order
        has_100 = true
      end

      out << up_to_100(cnt) if cnt > 0
      if has_100 or cnt > 0
        out << label
        @number -= cnt * order
      end
    }
    out << up_to_100(@number) if @number > 0
    out.join(" ")
  end

  def up_to_100(num)
    out = []
    has_tens = false

    if num >= 20
      how_many = num / 10
      out << TENS[how_many]
      num -= 10 * how_many
      has_tens = true
    end

    if num >= 10
      out << TEENS[num - 10]
      num = 0
    end

    if num > 0
      out << "-" if has_tens
      out << DIGITS[num]
    end

    out.join
  end
end
